snp_measurement/src/stage0.rs

```diff
@@ -87,3 +220,141 @@ pub fn load_stage0(stage0_rom_path: PathBuf) -> anyhow::Result<Stage0Info> {
     );
     Ok(Stage0Info::new(stage0_bytes))
 }
+
+/// Information about the pages specified in the firmware SEV metadata section entries.
+///
+/// See <https://github.com/tianocore/edk2/blob/fff6d81270b57ee786ea18ad74f43149b9f03494/OvmfPkg/ResetVector/X64/OvmfSevMetadata.asm>
+pub struct SevMetadataPageInfo {
+    pub start_address: PhysAddr,
+    pub page_count: usize,
+    pub page_type: PageType,
+}
+
+impl SevMetadataPageInfo {
+    fn parse(bytes: &[u8]) -> Self {
+        assert!(bytes.len() == SEV_METADATA_ENTRY_SIZE);
+        let mut base: u32 = 0;
+        base.as_bytes_mut().copy_from_slice(&bytes[0..4]);
+        let start_address = PhysAddr::new(base as u64);
+        assert_eq!(
+            start_address,
+            start_address.align_down(Size4KiB::SIZE),
+            "invalid SEV metadata entry base address"
+        );
+
+        let mut size: u32 = 0;
+        size.as_bytes_mut().copy_from_slice(&bytes[4..8]);
+        assert_eq!(
+            (size as u64) % Size4KiB::SIZE,
+            0,
+            "invalid SEV metadata entry size"
+        );
+        let page_count = (size as usize) / (Size4KiB::SIZE as usize);
+
+        let mut page_type: u32 = 0;
+        page_type.as_bytes_mut().copy_from_slice(&bytes[8..12]);
+        trace!(
+            "Metadata page entry: base: {}, size: {}, page_type: {}",
+            base,
+            size,
+            page_type
+        );
+        let page_type = SevMetadataPageType::from_repr(page_type)
+            .expect("invalid SEV metadata page type")
+            .into();
+
+        Self {
+            start_address,
+            page_count,
+            page_type,
+        }
+    }
+}
+
+/// The page types used in the firmware SEV metadata section entries.
+///
+/// See <https://github.com/tianocore/edk2/blob/fff6d81270b57ee786ea18ad74f43149b9f03494/OvmfPkg/ResetVector/X64/OvmfSevMetadata.asm>
+#[derive(Debug, FromRepr)]
+#[repr(u32)]
+enum SevMetadataPageType {
+    Invalid = 0,
+    Unmeasured = 1,
+    Secrets = 2,
+    Cpuid = 3,
+}
+
+impl From<SevMetadataPageType> for PageType {
+    fn from(value: SevMetadataPageType) -> Self {
+        match value {
+            SevMetadataPageType::Invalid => panic!("invalid SEV metadata page type"),
+            SevMetadataPageType::Cpuid => PageType::Cpuid,
+            SevMetadataPageType::Secrets => PageType::Secrets,
+            SevMetadataPageType::Unmeasured => PageType::Unmeasured,
+        }
+    }
+}
+
+/// The header of a guid table entry.
+///
+/// The table footer also has the same structure.
+struct GuidTableEntryHeader {
```

Comment:
Could this use zerocopy::FromBytes instead of manually parsing data?

Reply:
It complained about alignment mismatch with `repr(packed)` and a there was size mismatch due to the additional padding with `repr(C)`.